crates/pop-fork/src/rpc.rs

```diff
@@ -0,0 +1,346 @@
+pub struct ForkRpcClient {
+	legacy: LegacyRpcMethods<SubstrateConfig>,
+	endpoint: Url,
+}
+
+impl ForkRpcClient {
+	/// Connect to a live Polkadot-SDK chain.
+	///
+	/// # Arguments
+	/// * `endpoint` - WebSocket URL of the chain's RPC endpoint (e.g., `wss://rpc.polkadot.io`)
+	///
+	/// # Example
+	/// ```ignore
+	/// let client = ForkRpcClient::connect(&"wss://rpc.polkadot.io".parse()?).await?;
+	/// ```
+	pub async fn connect(endpoint: &Url) -> Result<Self, RpcClientError> {
+		let client = RpcClient::from_url(endpoint.as_str()).await.map_err(|e| {
+			RpcClientError::ConnectionFailed {
+				endpoint: endpoint.to_string(),
+				message: e.to_string(),
+			}
+		})?;
+
+		let legacy = LegacyRpcMethods::new(client);
+
+		Ok(Self { legacy, endpoint: endpoint.clone() })
+	}
+
+	/// Get the endpoint URL this client is connected to.
+	pub fn endpoint(&self) -> &Url {
+		&self.endpoint
+	}
+
+	/// Get the latest finalized block hash.
+	///
+	/// This is typically the starting point for forking - we fork from the latest
+	/// finalized state to ensure consistency.
+	pub async fn finalized_head(&self) -> Result<H256, RpcClientError> {
+		self.legacy
+			.chain_get_finalized_head()
+			.await
+			.map_err(|e| RpcClientError::RequestFailed {
+				method: methods::CHAIN_GET_FINALIZED_HEAD,
+				message: e.to_string(),
+			})
+	}
+
+	/// Get block header by hash.
+	///
+	/// Returns the header for the specified block, which contains the parent hash,
+	/// state root, extrinsics root, and digest.
+	pub async fn header(
+		&self,
+		hash: H256,
+	) -> Result<<SubstrateConfig as subxt::Config>::Header, RpcClientError> {
+		self.legacy
+			.chain_get_header(Some(hash))
+			.await
+			.map_err(|e| RpcClientError::RequestFailed {
+				method: methods::CHAIN_GET_HEADER,
+				message: e.to_string(),
+			})?
+			.ok_or_else(|| RpcClientError::InvalidResponse(format!("No header found for {hash:?}")))
+	}
+
+	/// Get a single storage value at a specific block.
+	///
+	/// # Arguments
+	/// * `key` - The storage key (raw bytes)
+	/// * `at` - The block hash to query state at
+	///
+	/// # Returns
+	/// * `Ok(Some(value))` - Storage exists with value
+	/// * `Ok(None)` - Storage key doesn't exist (empty)
+	/// * `Err(_)` - RPC error
+	pub async fn storage(&self, key: &[u8], at: H256) -> Result<Option<Vec<u8>>, RpcClientError> {
+		self.legacy.state_get_storage(key, Some(at)).await.map_err(|e| {
+			RpcClientError::RequestFailed {
+				method: methods::STATE_GET_STORAGE,
+				message: e.to_string(),
+			}
+		})
+	}
+
+	/// Get multiple storage values in a single batch request.
+	///
+	/// More efficient than multiple individual `storage()` calls when fetching
+	/// many keys at once.
+	///
+	/// # Arguments
+	/// * `keys` - Slice of storage keys to fetch
+	/// * `at` - The block hash to query state at
+	///
+	/// # Returns
+	/// A vector of optional values, in the same order as the input keys.
+	pub async fn storage_batch(
+		&self,
+		keys: &[Vec<u8>],
```

Comment:
I've changed the type of keys in https://github.com/r0gue-io/pop-cli/pull/840/changes#diff-efcd9fbee39f89e8765e88df7710ed298e46fed96a31c31e0672d5df66a95c2fR187

PR#840